File: pidex/core/cooldowns.py
import time

from pidex.core.constants import (
    DEFAULT_COOLDOWN_SSH_LOGIN,
    DEFAULT_COOLDOWN_SSH_LOGOUT,
    DEFAULT_COOLDOWN_SSH_BRUTEFORCE,
    DEFAULT_COOLDOWN_SUDO_USED,
    DEFAULT_COOLDOWN_CPU_HIGH,
    DEFAULT_COOLDOWN_CPU_RECOVERED,
    DEFAULT_COOLDOWN_TEMP_WARN,
    DEFAULT_COOLDOWN_TEMP_CRITICAL,
    DEFAULT_COOLDOWN_DISK_WARN,
    DEFAULT_COOLDOWN_DISK_CRITICAL,
    DEFAULT_COOLDOWN_RAM_HIGH,
    EVENT_SSH_LOGIN,
    EVENT_SSH_LOGOUT,
    EVENT_SSH_BRUTEFORCE,
    EVENT_SUDO_USED,
    EVENT_CPU_HIGH,
    EVENT_CPU_RECOVERED,
    EVENT_TEMP_WARN,
    EVENT_TEMP_CRITICAL,
    EVENT_DISK_WARN,
    EVENT_DISK_CRITICAL,
    EVENT_RAM_HIGH,
)
from pidex.core.event import Event


DEFAULT_COOLDOWNS: dict[str, float] = {
    EVENT_SSH_LOGIN: DEFAULT_COOLDOWN_SSH_LOGIN,
    EVENT_SSH_LOGOUT: DEFAULT_COOLDOWN_SSH_LOGOUT,
    EVENT_SSH_BRUTEFORCE: DEFAULT_COOLDOWN_SSH_BRUTEFORCE,
    EVENT_SUDO_USED: DEFAULT_COOLDOWN_SUDO_USED,
    EVENT_CPU_HIGH: DEFAULT_COOLDOWN_CPU_HIGH,
    EVENT_CPU_RECOVERED: DEFAULT_COOLDOWN_CPU_RECOVERED,
    EVENT_TEMP_WARN: DEFAULT_COOLDOWN_TEMP_WARN,
    EVENT_TEMP_CRITICAL: DEFAULT_COOLDOWN_TEMP_CRITICAL,
    EVENT_DISK_WARN: DEFAULT_COOLDOWN_DISK_WARN,
    EVENT_DISK_CRITICAL: DEFAULT_COOLDOWN_DISK_CRITICAL,
    EVENT_RAM_HIGH: DEFAULT_COOLDOWN_RAM_HIGH,
}


class CooldownManager:
    def __init__(self, overrides: dict[str, float] | None = None):
        self._cooled_until: dict[str, float] = {}
        self._durations: dict[str, float] = {**DEFAULT_COOLDOWNS}
        if overrides:
            self._durations.update(overrides)

    def is_allowed(self, event: Event) -> bool:
        now = time.time()
        deadline = self._cooled_until.get(event.event_type, 0)
        return now >= deadline

    def record(self, event: Event) -> None:
        duration = self._durations.get(event.event_type, 0)
        if duration > 0:
            self._cooled_until[event.event_type] = time.time() + duration

    def update_duration(self, event_type: str, seconds: float) -> None:
        self._durations[event_type] = seconds

File: internal/core/cooldowns.go
package core

import (
	"sync"
	"time"
)

var defaultCooldowns = map[string]float64{
	EventSSHLogin:     CooldownSSHLogin,
	EventSSHLogout:    CooldownSSHLogout,
	EventSSHBruteforce: CooldownSSHBruteforce,
	EventSudoUsed:     CooldownSudoUsed,
	EventCPUHigh:      CooldownCPUHigh,
	EventCPURecovered: CooldownCPURecovered,
	EventTempWarn:     CooldownTempWarn,
	EventTempCritical: CooldownTempCritical,
	EventDiskWarn:     CooldownDiskWarn,
	EventDiskCritical: CooldownDiskCritical,
	EventRAMHigh:      CooldownRAMHigh,
}

type CooldownManager struct {
	mu         sync.Mutex
	cooledUntil map[string]time.Time
	durations   map[string]time.Duration
}

func NewCooldownManager(overrides map[string]float64) *CooldownManager {
	durs := make(map[string]time.Duration, len(defaultCooldowns))
	for k, v := range defaultCooldowns {
		durs[k] = time.Duration(v) * time.Second
	}
	for k, v := range overrides {
		durs[k] = time.Duration(v) * time.Second
	}
	return &CooldownManager{
		cooledUntil: make(map[string]time.Time),
		durations:   durs,
	}
}

func (m *CooldownManager) IsAllowed(event Event) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	deadline, ok := m.cooledUntil[event.EventType]
	if !ok {
		return true
	}
	return time.Now().After(deadline)
}

func (m *CooldownManager) Record(event Event) {
	m.mu.Lock()
	defer m.mu.Unlock()
	dur, ok := m.durations[event.EventType]
	if !ok || dur == 0 {
		return
	}
	m.cooledUntil[event.EventType] = time.Now().Add(dur)
}

func (m *CooldownManager) UpdateDuration(eventType string, seconds float64) {
	m.mu.Lock()
	defer m.mu.Unlock()
	dur := time.Duration(seconds) * time.Second
	m.durations[eventType] = dur
	if seconds == 0 {
		delete(m.cooledUntil, eventType)
	}
}
